# Exercise 1: The Diet Problem

We  need to recommend supplement capsules for a person under your care. The daily requirements for five nutrients are as follow for solving through linear programming techniques.


## Definition of the problem

- 60 milligrams of Vitamin C
- 1000 milligrams of Calcium
- 18 milligrams of Iron
- 20 milligrams of Niacin
- 360 milligrams of Magnesium

### Available Suplements

You have two supplement options: **Vega Vita** and **Happy Health**.

- **Vega Vita** costs 20 cents per tablet and contains:
  - 20 mg of Vitamin C
  - 500 mg of Calcium
  - 9 mg of Iron
  - 2 mg of Niacin
  - 60 mg of Magnesium

- **Happy Health** costs 30 cents per tablet and contains:
  - 30 mg of Vitamin C
  - 250 mg of Calcium
  - 2 mg of Iron
  - 10 mg of Niacin
  - 90 mg of Magnesium



## Objective
Determine the minimum number of tablets from each type needed to meet daily nutritional requirements at the lowest cost.


In [686]:
!pip show pulp

Name: PuLP
Version: 2.9.0
Summary: PuLP is an LP modeler written in python. PuLP can generate MPS or LP files and call GLPK, COIN CLP/CBC, CPLEX, and GUROBI to solve linear problems.
Home-page: https://github.com/coin-or/pulp
Author: J.S. Roy and S.A. Mitchell and F. Peschiera
Author-email: pulp@stuartmitchell.com
License: 
Location: C:\Users\Luis\anaconda3\Lib\site-packages
Requires: 
Required-by: Pyfrontier


In [687]:
import pulp as pp

Let's start building the optimization model.

In [689]:
model = pp.LpProblem(name='diet-problem', # just the name
                     sense=pp.LpMinimize ) # type of problem

### Decision Variables
Define the following variables:
- \( x_1 \): Number of **Vega Vita** tablets.
- \( x_2 \): Number of **Happy Health** tablets.

We define the variables:

In [692]:
# how much vega?
x1 = pp.LpVariable(name="vega",  # just the name
                    lowBound=0,  # ensure non-negativity
                    cat='Continuous') # here: you accept decimal values

# how much happy?
x2 = pp.LpVariable(name="happy",
                 lowBound=0, 
                    
                 cat='Continuous')

### Non-Negativity Conditions
$$
x_1 \geq 0, \quad x_2 \geq 0
$$

This setup provides the full mathematical formulation of the problem, ready for solving through linear programming techniques.

### Objective Function (already defined)
Minimize the total daily cost of supplements:

$$
\text{Total Cost} = 0.20x_1 + 0.30x_2
$$


In [695]:
obj_func = x1*0.2 + x2*0.3

### Constraints
Each nutrient must meet or exceed the daily minimum requirement, leading to the following constraints:

1. **Vitamin C**: 
   $$
   20x_1 + 30x_2 \geq 60
   $$

2. **Calcium**: 
   $$
   500x_1 + 250x_2 \geq 1000
   $$

3. **Iron**: 
   $$
   9x_1 + 2x_2 \geq 18
   $$

4. **Niacin**: 
   $$
   2x_1 + 10x_2 \geq 20
   $$

5. **Magnesium**: 
   $$
   60x_1 + 90x_2 \geq 360
   $$


In [697]:
# SUBJECT TO:
C1 = pp.LpConstraint(name='Vitamin_C_Constraint',   
                     e= 20 * x1 + 30 * x2, rhs=60, 
                     sense=pp.LpConstraintGE)  # Vitamin C requirement >= 60 mg

C2 = pp.LpConstraint(name='Calcium_Constraint',     
                     e= 500 * x1 + 250 * x2, rhs=1000, 
                     sense=pp.LpConstraintGE)  # Calcium requirement >= 1000 mg

C3 = pp.LpConstraint(name='Iron_Constraint',        
                     e= 9 * x1 + 2 * x2, rhs=18, 
                     sense=pp.LpConstraintGE)  # Iron requirement >= 18 mg

C4 = pp.LpConstraint(name='Niacin_Constraint',      
                     e= 2 * x1 + 10 * x2, rhs=20, 
                     sense=pp.LpConstraintGE)  # Niacin requirement >= 20 mg

C5 = pp.LpConstraint(name='Magnesium_Constraint',   
                     e= 60 * x1 + 90 * x2, rhs=360, 
                     sense=pp.LpConstraintGE)  # Magnesium requirement >= 360 mg

### We build the model:

In [699]:
# Add constraints to the problem
model += obj_func
model += C1
model += C2
model += C3
model += C4
model += C5

### We solve the model:

In [701]:
solver_list = pp.listSolvers()
print(solver_list)

['GLPK_CMD', 'PYGLPK', 'CPLEX_CMD', 'CPLEX_PY', 'GUROBI', 'GUROBI_CMD', 'MOSEK', 'XPRESS', 'XPRESS', 'XPRESS_PY', 'PULP_CBC_CMD', 'COIN_CMD', 'COINMP_DLL', 'CHOCO_CMD', 'MIPCL_CMD', 'SCIP_CMD', 'FSCIP_CMD', 'SCIP_PY', 'HiGHS', 'HiGHS_CMD', 'COPT', 'COPT_DLL', 'COPT_CMD']


In [702]:
model.solve() #there is one solution

1

In [703]:
# Compile results into a dictionary
Results = {
    "Model Status": pp.LpStatus[model.status],
    "Optimal Solution": pp.value(model.objective)
}
Results.update({v.name: v.varValue for v in model.variables()})

# Convert results to a DataFrame for easy display
results_df = pd.DataFrame(list(Results.items()), columns=["Description", "Value"])
print(results_df)

        Description     Value
0      Model Status   Optimal
1  Optimal Solution       1.2
2             happy  3.130435
3              vega  1.304348


### Explanation

- **Model Status**: The model status is "Optimal," which means that the solution found meets the nutritional requirements at the lowest possible cost.
- **Optimal Solution (Total Cost)**: The minimum daily cost of meeting the nutrient requirements with the two supplements is $1.20.
- **Tablet Quantities**:
  - **happy**: 3.13 tablets of the "Happy" supplement are needed.
  - **vega**: 1.30 tablets of the "Vega" supplement are needed.

### Interpretation

These values indicate the precise number of tablets for each supplement required to meet the daily nutritional needs at the lowest cost. Although partial tablets are shown in the solution, they may need rounding based on practical constraints, while ensuring all nutritional requirements are still met.

# Exercise 2: Bus Driver Scheduling Problem

## Definition of the problem

The city transportation manager needs to schedule bus drivers to meet the following historical demand:

| Shift                 | Demand |
|-----------------------|--------|
| From midnight to 4 am | 4      |
| From 4 am to 8 am     | 8      |
| From 8 am to 12 pm    | 10     |
| From 12 pm to 4 pm    | 7      |
| From 4 pm to 8 pm     | 12     |
| From 8 pm to midnight | 4      |

### Objective
Minimize the number of drivers required at the start of each shift to meet the demand while adhering to the following conditions:
- Drivers work eight-hour shifts.
- Shifts start at times 0, 4, 8, 12, 16, or 20.
  - A driver starting at time 0 can work from 0 to 8.
  - A driver starting at time 20 will work the last four hours of one day and the first four hours of the next day.
- A driver does not need to drive continuously for the full eight hours; they may be idle for part of the shifn of the problem, ready for solving through linear programming techniques.


### Decision Variables
We define the following variables:
- \( x_0 \): Number of drivers starting at midnight (0:00).
- \( x_4 \): Number of drivers starting at 4 am.
- \( x_8 \): Number of drivers starting at 8 am.
- \( x_12 \): Number of drivers starting at 12 pm.
- \( x_16 \): Number of drivers starting at 4 pm.
- \( x_20 \): Number of drivers starting at 8 pm.

Let's start building the optimization model

In [709]:
import pulp as pp
import pandas as pd
model2 = pp.LpProblem("Driver_Scheduling_Minimization", pp.LpMinimize) #minimize

### Objective Function
Minimize the total number of drivers:

$$
\text{Total Drivers} = x_{0} + x_{4} + x_{8} + x_{12} + x_{16} + x_{20}
$$

We define the variables:

In [712]:

# Define decision variables
y0 = pp.LpVariable(name="drivers_midnight", lowBound=0, cat='Continuous')
y4 = pp.LpVariable(name="drivers_4am", lowBound=0, cat='Continuous')
y8 = pp.LpVariable(name="drivers_8am", lowBound=0, cat='Continuous')
y12 = pp.LpVariable(name="drivers_12pm", lowBound=0, cat='Continuous')
y16 = pp.LpVariable(name="drivers_4pm", lowBound=0, cat='Continuous')
y20 = pp.LpVariable(name="drivers_8pm", lowBound=0, cat='Continuous')

In [713]:

# Define the objective function (minimize total drivers)
obj_func2 = y0 + y4 + y8 + y12 + y16 + y20

### Non-Negativity Conditions (already defined)
$$
x_{0} \geq 0, \quad x_{4} \geq 0, \quad x_{8} \geq 0, \quad x_{12} \geq 0, \quad x_{16} \geq 0, \quad x_{20} \geq 0
$$


### Constraints
Each time period must meet or exceed the required demand:

1. **Midnight to 4 am**: 
   $$
   x_{0} + x_{20} \geq 4
   $$

2. **4 am to 8 am**: 
   $$
   x_{0} + x_{4} + x_{20} \geq 8
   $$

3. **8 am to 12 pm**: 
   $$
   x_{4} + x_{8} \geq 10
   $$

4. **12 pm to 4 pm**: 
   $$
   x_{8} + x_{12} \geq 7
   $$

5. **4 pm to 8 pm**: 
   $$
   x_{12} + x_{16} \geq 12
   $$

6. **8 pm to midnight**: 
   $$
   x_{16} + x_{20} \geq 4
   $$



In [716]:
# Define constraints with labels B1, B2, etc.
B1 = pp.LpConstraint(name='Midnight_to_4am_Constraint',   
                     e= y0 + y20, rhs=4, 
                     sense=pp.LpConstraintGE)  # Demand from midnight to 4 am >= 4 drivers

B2 = pp.LpConstraint(name='4am_to_8am_Constraint',     
                     e= y0 + y4 + y20, rhs=8, 
                     sense=pp.LpConstraintGE)  # Demand from 4 am to 8 am >= 8 drivers

B3 = pp.LpConstraint(name='8am_to_12pm_Constraint',        
                     e= y4 + y8, rhs=10, 
                     sense=pp.LpConstraintGE)  # Demand from 8 am to 12 pm >= 10 drivers

B4 = pp.LpConstraint(name='12pm_to_4pm_Constraint',      
                     e= y8 + y12, rhs=7, 
                     sense=pp.LpConstraintGE)  # Demand from 12 pm to 4 pm >= 7 drivers

B5 = pp.LpConstraint(name='4pm_to_8pm_Constraint',   
                     e= y12 + y16, rhs=12, 
                     sense=pp.LpConstraintGE)  # Demand from 4 pm to 8 pm >= 12 drivers

B6 = pp.LpConstraint(name='8pm_to_Midnight_Constraint',   
                     e= y16 + y20, rhs=4, 
                     sense=pp.LpConstraintGE)  # Demand from 8 pm to midnight >= 4 drivers


### We build the model:

In [718]:

# Add the objective function to model2
model2 += obj_func2, "Total_Drivers"

# Add constraints to the model
model2 += B1
model2 += B2
model2 += B3
model2 += B4
model2 += B5
model2 += B6

### We solve the model:

In [720]:
# Solve the model
model2.solve(pp.PULP_CBC_CMD())  # Using CBC solver explicitly

# Compile results into a dictionary
Results2 = {
    "Model Status": pp.LpStatus[model2.status],
    "Optimal Solution (Total Drivers)": pp.value(model2.objective)
}
Results2.update({v.name: v.varValue for v in model2.variables()})

# Convert results to a DataFrame for easy display
results2_df = pd.DataFrame(list(Results2.items()), columns=["Description", "Value"])
print(results2_df)

                        Description    Value
0                      Model Status  Optimal
1  Optimal Solution (Total Drivers)     26.0
2                      drivers_12pm      7.0
3                       drivers_4am     10.0
4                       drivers_4pm      5.0
5                       drivers_8am      0.0
6                       drivers_8pm      0.0
7                  drivers_midnight      4.0


### Explanation

- **Model Status**: The model status is "Optimal," indicating that a feasible and optimal solution was found for the problem.
- **Optimal Solution (Total Drivers)**: The total minimum number of drivers required to meet the demand across all shifts is 26.
- **Shift Assignments**:
  - **drivers_12pm**: 7 drivers are needed starting at 12 pm.
  - **drivers_4am**: 10 drivers are required to start at 4 am.
  - **drivers_4pm**: 5 drivers should start at 4 pm.
  - **drivers_8am**: No drivers are needed to start at 8 am.
  - **drivers_8pm**: No drivers are needed to start at 8 pm.
  - **drivers_midnight**: 4 drivers are needed starting at midnight.


# Exercise 3: Choosing a country for a Master Program

1. Prepare data file with the comparissons adn get the data:

In [898]:
import pandas as pd

# Lee el archivo Excel
master = "comparison.xlsx"

2. Open each sheet:

In [900]:
# opening the comparissons

pairwise_cost=pd.read_excel(master,sheet_name='cost', index_col=0)
pairwise_language=pd.read_excel(master,sheet_name='language', index_col=0)
pairwise_employ=pd.read_excel(master,sheet_name='employ', index_col=0)
pairwise_criteria=pd.read_excel(master,sheet_name='criteria', index_col=0)


we may want to check the structure:

In [902]:
pairwise_criteria

cost  language    employ
criteria                          
cost         1      0.25  0.166667
language     4      1.00  0.333333
employ       6      3.00  1.000000

3. Transform all matrices into pairwise comparissons:

In [904]:
import networkx as nx

G_cost = nx.from_pandas_adjacency(pairwise_cost,create_using=nx.MultiDiGraph())
G_language = nx.from_pandas_adjacency(pairwise_language, create_using=nx.MultiDiGraph())
G_employ = nx.from_pandas_adjacency(pairwise_employ, create_using=nx.MultiDiGraph())

In [906]:
# comparissons for cost as dict
cost_comparisons ={(e[0],e[1]):e[2]['weight'] for e in G_cost.edges(data=True) if e[0]!= e[1]}
language_comparisons = {(e[0], e[1]): e[2]['weight'] for e in G_language.edges(data=True) if e[0] != e[1]}
employ_comparisons = {(e[0], e[1]): e[2]['weight'] for e in G_employ.edges(data=True) if e[0] != e[1]}

In [908]:
# take a look
[cost_comparisons, language_comparisons,employ_comparisons,]

[{('Brazil', 'Spain'): 0.3333333333333333,
  ('Brazil', 'Germany'): 0.2,
  ('Brazil', 'USA'): 0.1428571428571428,
  ('Spain', 'Brazil'): 3.0,
  ('Spain', 'Germany'): 0.25,
  ('Spain', 'USA'): 0.1666666666666667,
  ('Germany', 'Brazil'): 5.0,
  ('Germany', 'Spain'): 4.0,
  ('Germany', 'USA'): 0.3333333333333333,
  ('USA', 'Brazil'): 7.0,
  ('USA', 'Spain'): 6.0,
  ('USA', 'Germany'): 3.0},
 {('Brazil', 'Spain'): 0.5,
  ('Brazil', 'Germany'): 0.1428571428571428,
  ('Brazil', 'USA'): 0.25,
  ('Spain', 'Brazil'): 2.0,
  ('Spain', 'Germany'): 0.1111111111111111,
  ('Spain', 'USA'): 0.1666666666666667,
  ('Germany', 'Brazil'): 7.0,
  ('Germany', 'Spain'): 9.0,
  ('Germany', 'USA'): 0.3333333333333333,
  ('USA', 'Brazil'): 4.0,
  ('USA', 'Spain'): 6.0,
  ('USA', 'Germany'): 3.0},
 {('Brazil', 'Spain'): 2.0,
  ('Brazil', 'Germany'): 5.0,
  ('Brazil', 'USA'): 7.0,
  ('Spain', 'Brazil'): 0.5,
  ('Spain', 'Germany'): 3.0,
  ('Spain', 'USA'): 5.0,
  ('Germany', 'Brazil'): 0.2,
  ('Germany', 'Spain

In [910]:
# now the criteria

G_CRIT = nx.from_pandas_adjacency(pairwise_criteria,create_using=nx.MultiDiGraph())
criteria_comparisons ={(e[0],e[1]):e[2]['weight'] for e in G_CRIT.edges(data=True) if e[0]!= e[1]}
criteria_comparisons

{('cost', 'language'): 0.25,
 ('cost', 'employ'): 0.1666666666666667,
 ('language', 'cost'): 4.0,
 ('language', 'employ'): 0.3333333333333333,
 ('employ', 'cost'): 6.0,
 ('employ', 'language'): 3.0}

4. Apply the Algorithm

In [ ]:
## install
!pip install ahpy

In [912]:
# input each comparisson

import ahpy

cost = ahpy.Compare('cost', cost_comparisons, precision=3, random_index='saaty')
language = ahpy.Compare('language', language_comparisons, precision=3, random_index='saaty')
employ = ahpy.Compare('employ', employ_comparisons, precision=3, random_index='saaty')
criteria = ahpy.Compare('criteria', criteria_comparisons, precision=3, random_index='saaty')

5. Create hierarchy:

In [914]:
criteria.add_children([cost, language, employ])

6. See results

In [916]:
print(criteria.target_weights)

{'Brazil': 0.351, 'USA': 0.261, 'Spain': 0.223, 'Germany': 0.165}


### Interpretación de Resultados de `criteria.target_weights`

Los resultados de `criteria.target_weights` muestran las preferencias finales por país al combinar tres criterios: **costo de vida** (`cost`), **dificultad del idioma** (`language`) y **posibilidades de empleo** (`employ`). Estos valores representan la prioridad relativa de cada país al considerar estos tres factores en conjunto.

#### Interpretación de los valores obtenidos:

- **Brasil (0.351)**: Es el país más preferido según los criterios de costo de vida, facilidad del idioma y posibilidad de empleo, con una prioridad del 35.1%. Esto sugiere que, en general, Brasil ofrece la combinación más favorable de estos factores para la decisión que se está evaluando.

- **EE. UU. (0.261)**: Le sigue en preferencia, con una ponderación del 26.1%. Aunque no es tan favorable como Brasil en la combinación de los tres criterios, sigue siendo una opción importante en la evaluación.

- **España (0.223)**: Con una prioridad del 22.3%, España es menos preferida que Brasil y EE. UU., pero sigue siendo una opción viable, especialmente si algunos factores (como el idioma) son favorables para el contexto.

- **Alemania (0.165)**: Es el país menos preferido en esta evaluación, con una prioridad de 16.5%. Esto indica que, considerando el costo de vida, la dificultad del idioma y las oportunidades de empleo, Alemania ofrece la combinación menos favorable en comparación con los otros paEntonces
#### Resumen:
Este análisis sugiere que, al equilibrar los factores de **costo de vida**, **dificultad del idioma** y **empleabilidad**, Brasil es la mejor opción, seguido por EE. UU. y España, mientras que Alemania es la opción menos favorable. Este tipo de resultado es útil para tomar decisiones sobre dónde establecerse o enfocar oportunidades, basándose en las prioridades establecidas en cada criterio.


7. Assess consistency

In [919]:
## We should review comparissons if greater than 0.1!
[(val.name,val.consistency_ratio) for val in [cost, language, employ, criteria]]

[('cost', 0.065), ('language', 0.174), ('employ', 0.167), ('criteria', 0.052)]

### Interpretación de los Índices de Consistencia de Comparaciones

La salida obtenida de los índices de consistencia muestra qué tan consistentes son las comparaciones de cada criterio en el proceso de análisis jerárquico (AHP). Un índice de consistencia superior a 0.1 indica que las comparaciones pueden no ser consistentes y, por lo tanto, podría ser necesario revisarlas para mejorar la coherencia de los resultados.

#### Resultados de Consistencia:

- **Cost (`cost`, 0.065)**: La consistencia es adecuada, con un valor de 0.065, que está por debajo del umbral de 0.1. Esto indica que las comparaciones de costo de vida son consistentes y no requieren revisión.

- **Language (`language`, 0.174)**: El índice de consistencia es 0.174, lo cual supera el umbral de 0.1. Esto sugiere que las comparaciones en términos de dificultad del idioma no son completamente consistentes y podrían necesitar revisión para garantizar una mejor precisión en el análisis.

- **Employ (`employ`, 0.167)**: Similar al criterio anterior, el índice de consistencia es 0.167, que también excede el umbral. Esto indica una falta de consistencia en las comparaciones de posibilidades de empleo, lo que sugiere la necesidad de revisar estas comparaciones.

- **Criteria (`criteria`, 0.052)**: El índice de consistencia es 0.052, lo que está bien por debajo del umbral. Esto indica que las comparaciones entre los criterios generales son consistentes y no requieren ajuste.

#### Conclusión

Los criterios de **dificultad del idioma (`language`)** y **posibilidades de empleo (`employ`)** tienen índices de consistencia superiores a 0.1, lo que indica una falta de consistencia en las comparaciones. Sería recomendable revisar estos dos criterios para asegurarse de que las evaluaciones sean más coherentes, lo cual puede mejorar la precisión del análisis AHP. Los criterios **costo de vida (`cost`)** y **comparación general de criterios (`criteria`)** son consistentes y no requieren cambios.


# Exercise 4: Efficiency in Public sectorency.

In [987]:
import pandas as pd

# Lee el archivo especificando el delimitador como ";"
muni = pd.read_csv("2023.csv", sep=";", low_memory=False)

#  
muni.head()


Año  idmunici  ccdd  ccpp  ccdi  Ubigeo Departamento    Provincia  \
0  2023     10101     1     1     1   10101     AMAZONAS  CHACHAPOYAS   
1  2023     10102     1     1     2   10102     AMAZONAS  CHACHAPOYAS   
2  2023     10103     1     1     3   10103     AMAZONAS  CHACHAPOYAS   
3  2023     10104     1     1     4   10104     AMAZONAS  CHACHAPOYAS   
4  2023     10105     1     1     5   10105     AMAZONAS  CHACHAPOYAS   

      Distrito  Tipomuni  ...  C97_2_2_3_6  C97_2_2_3_7 C97_2_2_3_8  \
0  CHACHAPOYAS         1  ...          NaN          NaN         NaN   
1     ASUNCION         2  ...            0            0           0   
2       BALSAS         2  ...            0            0           0   
3        CHETO         2  ...            0            0           0   
4    CHILIQUIN         2  ...            0            0           0   

  C97_2_2_3_9 C97_2_2_3_10 C97_2_2_3_11  C97_2_2_4 C97_2_3 C97_2_4 C97_3  
0         NaN          NaN          NaN        NaN     NaN     NaN   NaN  
1           0       108292            0  143238.16       0       0     0  
2           0    176714.76            0      97960       0       0     0  
3           0        17000            0     147000       0       0     0  
4           0            0            0   453742.7       0       0     0  

[5 rows x 1383 columns]

In [999]:
# Filtrar el DataFrame para quedarte con los registros donde el Distrito es 'Chachapoyas'
muni_chachapoyas = muni[muni['Departamento'] == 'AMAZONAS']
muni_chachapoyas

Año  idmunici  ccdd  ccpp  ccdi  Ubigeo Departamento    Provincia  \
0   2023     10101     1     1     1   10101     AMAZONAS  CHACHAPOYAS   
1   2023     10102     1     1     2   10102     AMAZONAS  CHACHAPOYAS   
2   2023     10103     1     1     3   10103     AMAZONAS  CHACHAPOYAS   
3   2023     10104     1     1     4   10104     AMAZONAS  CHACHAPOYAS   
4   2023     10105     1     1     5   10105     AMAZONAS  CHACHAPOYAS   
..   ...       ...   ...   ...   ...     ...          ...          ...   
79  2023     10703     1     7     3   10703     AMAZONAS    UTCUBAMBA   
80  2023     10704     1     7     4   10704     AMAZONAS    UTCUBAMBA   
81  2023     10705     1     7     5   10705     AMAZONAS    UTCUBAMBA   
82  2023     10706     1     7     6   10706     AMAZONAS    UTCUBAMBA   
83  2023     10707     1     7     7   10707     AMAZONAS    UTCUBAMBA   

        Distrito  Tipomuni  ...  C97_2_2_3_6  C97_2_2_3_7 C97_2_2_3_8  \
0    CHACHAPOYAS         1  ...          NaN          NaN         NaN   
1       ASUNCION         2  ...            0            0           0   
2         BALSAS         2  ...            0            0           0   
3          CHETO         2  ...            0            0           0   
4      CHILIQUIN         2  ...            0            0           0   
..           ...       ...  ...          ...          ...         ...   
79         CUMBA         2  ...            0   3655088.66      307203   
80    EL MILAGRO         2  ...            0   2927897.73   775807.43   
81       JAMALCA         2  ...            0    112296.16           0   
82  LONYA GRANDE         2  ...            0   2528169.66   246456.79   
83         YAMON         2  ...            0    6219758.7           0   

   C97_2_2_3_9 C97_2_2_3_10 C97_2_2_3_11  C97_2_2_4 C97_2_3 C97_2_4 C97_3  
0          NaN          NaN          NaN        NaN     NaN     NaN   NaN  
1            0       108292            0  143238.16       0       0     0  
2            0    176714.76            0      97960       0       0     0  
3            0        17000            0     147000       0       0     0  
4            0            0            0   453742.7       0       0     0  
..         ...          ...          ...        ...     ...     ...   ...  
79           0    881395.52            0  1135532.1       0       0     0  
80           0    870102.88            0     366275       0       0     0  
81           0    955086.23    356637.82  734360.89       0       0     0  
82   430715.56   1320616.35            0          0       0       0     0  
83           0    182098.78            0  566828.27       0       0     0  

[84 rows x 1383 columns]

The first three variables (Aircraft,Fuel,Employee) represent inputs and the last two ones represent outputs. If that is so, there should be a way to compute some measure of efficiency: the ratio output/input.

Let's compute some ratios:

In [ ]:
# ratio passenger employee:
airline['rate_ClientsByEmployee']=(airline.Passenger/airline.Employee)
airline['rate_CargoByFleet']=(airline.Freight/airline.Aircraft)

Let me plot those ratios:

In [ ]:
import altair as alt

points = alt.Chart(airline).mark_point().encode(
    x='rate_ClientsByEmployee:Q',
    y='rate_CargoByFleet:Q'
)

text = points.mark_text(
    align='right',
    baseline='middle',
    dx=-7
).encode(
    text='name'
).interactive()

points + text

Which one is more efficient? As you see, one airline might not be good in both ratios:

In [ ]:
airline[['name','rate_ClientsByEmployee','rate_CargoByFleet']].sort_values(by='rate_ClientsByEmployee',ascending=False).head()

In [ ]:
airline[['name','rate_ClientsByEmployee','rate_CargoByFleet']].sort_values(by='rate_CargoByFleet',ascending=False).head()

Let me show you the envelope:

In [ ]:
Best_ClientsByEmployee=airline.rate_ClientsByEmployee.idxmax()
Best_CargoByFleet=airline.rate_CargoByFleet.idxmax()

frontier1=airline.loc[Best_ClientsByEmployee,['rate_ClientsByEmployee','rate_CargoByFleet']].to_list()
frontier2=airline.loc[Best_CargoByFleet,['rate_ClientsByEmployee','rate_CargoByFleet']].to_list()

#parallels
frontier1v=[frontier1[0],0]
frontier2h=[0,frontier2[1]]

#then
envelope=pd.DataFrame([frontier2h,frontier2,frontier1,frontier1v],columns=['x','y'])
envelope

In [ ]:
points + text + alt.Chart(envelope).mark_line(color='red').encode(
    x='x',
    y='y',
)

The presence of several units (DMUs), several inputs, and several outputs makes it difficult to judge who is doing better. This an optimization problem that may be carried out using Pyfrontier:

In [ ]:
## installation
!pip install Pyfrontier

In [ ]:
airlineInput=airline.columns[1:4].to_list()
airlineOutput=airline.columns[4:6].to_list()

In [ ]:
from Pyfrontier.frontier_model import EnvelopDEA

dea_air_vrs_in = EnvelopDEA("VRS", "in")
dea_air_vrs_in.fit(
    inputs=airline[airlineInput].to_numpy(),
    outputs=airline[airlineOutput].to_numpy()
)

In [ ]:
airline['vrs_in']=[r.score for r in dea_air_vrs_in.result]
airline.set_index(airline.name,inplace=True)
airline['vrs_in']

In [ ]:
!pip install py4etrics


In [ ]:
import numpy as np # linear algebra
from py4etrics import tobit
# import statsmodels.api as sm

airline['censored'] =np.where(airline['vrs_in']==1, 1, 0)
cens = airline['censored']
endog = airline.loc[:,'vrs_in']
exog = airline.loc[:,'Aircraft':'Freight']

tobit_res = tobit.Tobit(endog, exog, cens,right=1).fit()
tobit_res.summary()

# Use github para almacenar, publicar y presentar su trabajo

Enlace al repositorio de la tarea 3: https://github.com/luispachecoc/optimization

Enlace a GitHub Pages: https://luispachecoc.github.io/optimization/